# 00 - Dataset introduction

Section **3.1 Introduction** of the report.

## Brief

Describe the dataset: source, what each row/column represents, define the target variable and the type of problem (classification or regression).

- Origin and source of the dataset.
- Description of rows (observational unit) and columns (features).
- Target variable.
- Problem type: classification / regression.

## Source

Dataset: [Wine Reviews](https://www.kaggle.com/datasets/zynicide/wine-reviews/data) — scraped from WineEnthusiast magazine by Zack Thoutt (Kaggle).

Three files are available in `data/raw/`:

| File | Format | Rows | Columns |
|---|---|---|---|
| `winemag-data-130k-v2.csv` | CSV | 129,971 | 13 |
| `winemag-data_first150k.csv` | CSV | 150,930 | 10 |
| `winemag-data-130k-v2.json` | JSON | 129,971 | 13 |

The **v2 CSV** (and its JSON mirror) is the primary file for this project — it is the most complete version, containing taster metadata not present in the 150k file.

In [1]:
import json

import numpy as np
import pandas as pd
from IPython.display import display

from diplo_mod_1.constants import RAW

## winemag-data-130k-v2.csv (primary)

129,971 wine reviews. Each row is one review written by a WineEnthusiast taster.

In [2]:
df_v2 = pd.read_csv(RAW / "winemag-data-130k-v2.csv", index_col=0)
print(f"Shape: {df_v2.shape}")
print(f"\nDtypes:\n{df_v2.dtypes}")
print(f"\nMissing values:\n{df_v2.isnull().sum()}")
df_v2.describe(include="all")

Shape: (129971, 13)

Dtypes:
country                   object
description               object
designation               object
points                     int64
price                    float64
province                  object
region_1                  object
region_2                  object
taster_name               object
taster_twitter_handle     object
title                     object
variety                   object
winery                    object
dtype: object

Missing values:
country                     63
description                  0
designation              37465
points                       0
price                     8996
province                    63
region_1                 21247
region_2                 79460
taster_name              26244
taster_twitter_handle    31213
title                        0
variety                      1
winery                       0
dtype: int64


,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
count,129908,129971,92506,129971.000000,120975.000000,129908,108724,50511,103727,98758,129971,129970,129971
unique,43,119955,37979,NaN,NaN,425,1229,17,19,15,118840,707,16757
top,US,"Seductively tart in lemon pith, cranberry and ...",Reserve,NaN,NaN,California,Napa Valley,Central Coast,Roger Voss,@vossroger,Gloria Ferrer NV Sonoma Brut Sparkling (Sonoma...,Pinot Noir,Wines & Winemakers
freq,54504,3,2009,NaN,NaN,36247,4480,11065,25514,25514,11,13272,222
mean,NaN,NaN,NaN,88.447138,35.363389,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,3.039730,41.022218,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,80.000000,4.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,86.000000,17.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,88.000000,25.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,91.000000,42.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Column descriptions

| Column | Type | Description | Missing |
|---|---|---|---|
| `country` | categorical | Country where the wine was produced | 63 |
| `description` | text | Taster's written review | 0 |
| `designation` | categorical | Vineyard sub-label within the winery | 37,465 |
| `points` | int | Score assigned by the taster (80–100) | 0 |
| `price` | float | Bottle price in USD | 8,996 |
| `province` | categorical | Province or state of origin | 63 |
| `region_1` | categorical | Specific wine-growing region | 21,247 |
| `region_2` | categorical | More specific sub-region | 79,460 |
| `taster_name` | categorical | Name of the WineEnthusiast reviewer | 26,244 |
| `taster_twitter_handle` | categorical | Twitter handle of the reviewer | 31,213 |
| `title` | text | Full wine label (includes vintage year) | 0 |
| `variety` | categorical | Grape variety (e.g. Pinot Noir, Chardonnay) | 1 |
| `winery` | categorical | Name of the producing winery | 0 |

**Key stats:**
- `points`: mean 88.4, std 3.0, range 80–100
- `price`: mean $35.4, std $41.0, range $4–$3,300 (heavy right skew)
- `country`: 43 unique values — US dominates with 54,504 reviews
- `variety`: 707 unique grape varieties

## winemag-data_first150k.csv (older version)

150,930 wine reviews. An earlier scrape — larger but missing the taster columns present in v2.

In [3]:
df_150k = pd.read_csv(RAW / "winemag-data_first150k.csv", index_col=0)
print(f"Shape: {df_150k.shape}")
print(f"\nDtypes:\n{df_150k.dtypes}")
print(f"\nMissing values:\n{df_150k.isnull().sum()}")
df_150k.describe(include="all")

Shape: (150930, 10)

Dtypes:
country         object
description     object
designation     object
points           int64
price          float64
province        object
region_1        object
region_2        object
variety         object
winery          object
dtype: object

Missing values:
country            5
description        0
designation    45735
points             0
price          13695
province           5
region_1       25060
region_2       89977
variety            0
winery             0
dtype: int64


,country,description,designation,points,price,province,region_1,region_2,variety,winery
count,150925,150930,105195,150930.000000,137235.000000,150925,125870,60953,150930,150930
unique,48,97821,30621,NaN,NaN,455,1236,18,632,14810
top,US,A little bit funky and unsettled when you pop ...,Reserve,NaN,NaN,California,Napa Valley,Central Coast,Chardonnay,Williams Selyem
freq,62397,6,2752,NaN,NaN,44508,6209,13057,14482,374
mean,NaN,NaN,NaN,87.888418,33.131482,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,3.222392,36.322536,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,80.000000,4.000000,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,86.000000,16.000000,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,88.000000,24.000000,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,90.000000,40.000000,NaN,NaN,NaN,NaN,NaN


### Differences from v2

| Aspect | v2 (130k) | first150k |
|---|---|---|
| Rows | 129,971 | 150,930 |
| Columns | 13 | 10 |
| `taster_name` | yes | **no** |
| `taster_twitter_handle` | yes | **no** |
| `title` | yes | **no** |
| `points` mean | 88.4 | 87.9 |
| `price` mean | $35.4 | $33.1 |
| `variety` unique | 707 | 632 |

This file will **not** be used for modelling — it is kept for reference only.

### v2 vs first150k comparison

Four checks to determine whether the 150k dataset can enrich the v2:

1. **Column diff** — which columns are shared vs exclusive to each file.
2. **Row overlap via `description`** — inner join on review text to count matching rows.
3. **Exclusive rows** — rows in 150k with no match in v2 (potential new data).
4. **Value mismatches** — for matched rows, do shared column values agree?

In [4]:
# 1. Column diff
cols_v2 = set(df_v2.columns)
cols_150k = set(df_150k.columns)
shared_cols = cols_v2 & cols_150k

print("=== 1. Column diff ===")
print(f"Shared ({len(shared_cols)}):       {sorted(shared_cols)}")
print(f"Only in v2 ({len(cols_v2 - cols_150k)}):   {sorted(cols_v2 - cols_150k)}")
print(f"Only in 150k ({len(cols_150k - cols_v2)}): {sorted(cols_150k - cols_v2)}")

# 2. Row overlap via description
matched = df_v2[["description"]].merge(df_150k[["description"]], on="description", how="inner")

print("\n=== 2. Row overlap via description ===")
print(f"Rows in v2:           {len(df_v2):>7}")
print(f"Rows in 150k:         {len(df_150k):>7}")
print(f"Matched (inner join): {len(matched):>7}")
print(f"Match rate v2:        {len(matched) / len(df_v2) * 100:.1f}%")
print(f"Match rate 150k:      {len(matched) / len(df_150k) * 100:.1f}%")

# 3. Exclusive rows in 150k
n_exclusive = (~df_150k["description"].isin(df_v2["description"])).sum()
print("\n=== 3. Exclusive rows in 150k ===")
print(f"Rows with no match in v2: {n_exclusive:>7}")

# Duplicate descriptions (context for check 4)
print("\n=== Duplicate descriptions ===")
print(f"v2   duplicates: {df_v2['description'].duplicated().sum()}")
print(f"150k duplicates: {df_150k['description'].duplicated().sum()}")

=== 1. Column diff ===
Shared (10):       ['country', 'description', 'designation', 'points', 'price', 'province', 'region_1', 'region_2', 'variety', 'winery']
Only in v2 (3):   ['taster_name', 'taster_twitter_handle', 'title']
Only in 150k (0): []

=== 2. Row overlap via description ===
Rows in v2:            129971
Rows in 150k:          150930
Matched (inner join):   73571
Match rate v2:        56.6%
Match rate 150k:      48.7%

=== 3. Exclusive rows in 150k ===
Rows with no match in v2:   82041

=== Duplicate descriptions ===
v2   duplicates: 10016
150k duplicates: 53109


In [5]:
# 4. Value mismatches on shared columns for matched rows
compare_cols = sorted(shared_cols - {"description"})
merged = df_v2[list(shared_cols)].merge(
    df_150k[list(shared_cols)],
    on="description",
    how="inner",
    suffixes=("_v2", "_150k"),
)

print(f"=== 4. Value mismatches (shared cols, {len(merged)} matched rows) ===")
any_mismatch = False
for col in compare_cols:
    left, right = f"{col}_v2", f"{col}_150k"
    diff = (merged[left] != merged[right]) & merged[left].notna() & merged[right].notna()
    n = diff.sum()
    if n:
        any_mismatch = True
        print(f"\n  {col}: {n} mismatches")
        display(merged[[left, right]][diff].head(3))

if not any_mismatch:
    print("  No mismatches — all shared column values agree on matched rows")

=== 4. Value mismatches (shared cols, 73571 matched rows) ===

  country: 6 mismatches


,country_v2,country_150k
12846,Argentina,France
12848,Argentina,France
12850,Argentina,France



  designation: 24 mismatches


,designation_v2,designation_150k
4321,Les Clissages d'Or Sur Lie,Les Clissages d'Or sur lie
4322,Les Clissages d'Or Sur Lie,Les Clissages d'Or sur lie
6080,Woods Vineyard Rosé of,800 Vines Jenna's Vineyard Rosé of



  points: 22 mismatches


,points_v2,points_150k
6080,91,89
17341,88,90
22911,87,85



  price: 34 mismatches


,price_v2,price_150k
471,28.0,16.0
472,28.0,16.0
6080,29.0,25.0



  province: 12 mismatches


,province_v2,province_150k
12846,Mendoza Province,Burgundy
12848,Mendoza Province,Burgundy
12850,Mendoza Province,Burgundy



  region_1: 164 mismatches


,region_1_v2,region_1_150k
118,Willamette Valley,Yamhill-Carlton District
1624,Willamette Valley,Yamhill-Carlton District
1771,Spring Mountain District,Napa



  region_2: 14 mismatches


,region_2_v2,region_2_150k
371,Napa-Sonoma,North Coast
372,Napa-Sonoma,North Coast
24014,Napa-Sonoma,North Coast



  variety: 11 mismatches


,variety_v2,variety_150k
8755,Muscat Blanc à Petits Grains,Muscat Blanc à Petit Grain
12846,Malbec,Pinot Noir
12848,Malbec,Pinot Noir



  winery: 255 mismatches


,winery_v2,winery_150k
725,Nefarious,Nefarious Cellars
943,Gerovassiliou,Domaine Gerovassiliou
1080,Château de Pizay,Chateau de Pizay


**Conclusion:** the 150k dataset is not worth merging into v2.

- It adds **no new columns** — v2 already has a strict superset.
- Only **56.6% of v2 rows** match by description text, and the description key is unreliable (53k duplicate descriptions in 150k).
- Even "matched" rows disagree on key values (`country`, `points`, `price`, `winery`), indicating these are **two independent scrapes** with real data differences, not just formatting variants.
- The 82k exclusive rows in 150k come without `title`, `taster_name`, or `taster_twitter_handle`.

**Decision: use `winemag-data-130k-v2.csv` as the sole data source for all subsequent notebooks.**

## winemag-data-130k-v2.json

Same 129,971 records and 13 fields as the v2 CSV — just a different format. Kept as an alternative ingestion path.

In [6]:
with open(RAW / "winemag-data-130k-v2.json") as f:
    data_json = json.load(f)

print(f"Type: {type(data_json)}")
print(f"Records: {len(data_json)}")
print(f"Fields: {list(data_json[0].keys())}")

Type: <class 'list'>
Records: 129971
Fields: ['points', 'title', 'description', 'taster_name', 'taster_twitter_handle', 'price', 'designation', 'variety', 'region_1', 'region_2', 'province', 'country', 'winery']


### CSV vs JSON equivalence check

Both v2 files claim to carry the same 129,971 records. We verify this by:

1. Aligning column order (JSON field order differs from CSV).
2. Resolving dtype mismatches before comparing (the JSON parser loads `points` as `object` instead of `int64`).
3. Using `DataFrame.equals()`, which treats `NaN == NaN` as `True` — essential given the many nulls in this dataset.

In [7]:
df_json = pd.DataFrame(data_json)

# Align column order to match CSV
df_json = df_json[df_v2.columns]

# Report any dtype mismatches
dtype_mismatches = {
    col: (df_v2[col].dtype, df_json[col].dtype)
    for col in df_v2.columns
    if df_v2[col].dtype != df_json[col].dtype
}
print("Dtype mismatches:", dtype_mismatches)

# Cast JSON columns to match CSV dtypes
for col, (csv_dtype, _) in dtype_mismatches.items():
    df_json[col] = df_json[col].astype(csv_dtype)

# Compare (reset index so positional alignment is explicit)
are_identical = df_v2.reset_index(drop=True).equals(df_json.reset_index(drop=True))
print(f"\nFiles identical after dtype alignment: {are_identical}")

Dtype mismatches: {'points': (dtype('int64'), dtype('O'))}

Files identical after dtype alignment: True


In [8]:
rng = np.random.default_rng(42)
sample_idx = sorted(rng.choice(len(df_v2), size=5, replace=False))

csv_sample = df_v2.reset_index(drop=True).iloc[sample_idx]
json_sample = df_json.iloc[sample_idx]

print("=== CSV rows ===")
display(csv_sample)

print("=== JSON rows (same positions) ===")
display(json_sample)

print("=== Diff (empty means identical) ===")
display(csv_sample.reset_index(drop=True).compare(json_sample.reset_index(drop=True)))

=== CSV rows ===


,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
11599,Italy,"This sweet, sparkling Rosato opens with fresh ...",Birbet,83,17.0,Piedmont,Piedmont,NaN,NaN,NaN,Demarie NV Birbet Brachetto (Piedmont),Brachetto,Demarie
56279,France,"An intensely fruity Champagne, made for the Ts...",Tsarine Premier Cru Brut,89,60.0,Champagne,Champagne,NaN,Roger Voss,@vossroger,Chanoine NV Tsarine Premier Cru Brut (Champagne),Champagne Blend,Chanoine
57041,Italy,This pretty 50-50 blend of Chardonnay and Pino...,Dosaggio Zero,90,35.0,Lombardy,Franciacorta,NaN,NaN,NaN,Contadi Castaldi 2006 Dosaggio Zero Sparkling ...,Sparkling Blend,Contadi Castaldi
85074,France,Beautifully perfumed in a relatively light sty...,Santenay-Beauregard Premier Cru,89,60.0,Burgundy,Santenay,NaN,Roger Voss,@vossroger,Chanson Père et Fils 2008 Santenay-Beauregard ...,Pinot Noir,Chanson Père et Fils
100589,Portugal,"The local variety in Palmela is Castelão, plan...",Pegos Claros,85,10.0,Palmela,NaN,NaN,Roger Voss,@vossroger,Companhia das Quintas 2002 Pegos Claros Castel...,Castelão,Companhia das Quintas


=== JSON rows (same positions) ===


,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
11599,Italy,"This sweet, sparkling Rosato opens with fresh ...",Birbet,83,17.0,Piedmont,Piedmont,None,None,None,Demarie NV Birbet Brachetto (Piedmont),Brachetto,Demarie
56279,France,"An intensely fruity Champagne, made for the Ts...",Tsarine Premier Cru Brut,89,60.0,Champagne,Champagne,None,Roger Voss,@vossroger,Chanoine NV Tsarine Premier Cru Brut (Champagne),Champagne Blend,Chanoine
57041,Italy,This pretty 50-50 blend of Chardonnay and Pino...,Dosaggio Zero,90,35.0,Lombardy,Franciacorta,None,None,None,Contadi Castaldi 2006 Dosaggio Zero Sparkling ...,Sparkling Blend,Contadi Castaldi
85074,France,Beautifully perfumed in a relatively light sty...,Santenay-Beauregard Premier Cru,89,60.0,Burgundy,Santenay,None,Roger Voss,@vossroger,Chanson Père et Fils 2008 Santenay-Beauregard ...,Pinot Noir,Chanson Père et Fils
100589,Portugal,"The local variety in Palmela is Castelão, plan...",Pegos Claros,85,10.0,Palmela,None,None,Roger Voss,@vossroger,Companhia das Quintas 2002 Pegos Claros Castel...,Castelão,Companhia das Quintas


=== Diff (empty means identical) ===


Empty DataFrame
Columns: []
Index: []

**Result:** the CSV and JSON are perfect mirrors. The only difference was a dtype artifact — `points` loads as `object` from JSON vs `int64` from CSV — not a data discrepancy.

The CSV (`winemag-data-130k-v2.csv`) will be used as the single ingestion source for all subsequent notebooks.